### PR TITLE
feat(generic-metrics): Add migrations to allow sampling for counters

### PR DESCRIPTION
**Pull request by john-z-yang**
# Overview

Adds a set of migrations to allow materialization to account for sampled counter values.
Specifically
- Adds a new column that contains `sampling_weight`, i.e. `sampling factor^(-1)`, AKA how many actual row it is worth in the raw table
- Adds a new `value_weighted` column that contains the aggregated sum taking the sampling weight into consideration
- Adds a new materialized view to manifest the columns above.

# Local testing

<details>
  <summary>Click to expand</summary>
  
  ### Run the migrations

```bash
make apply-migrations
```

  ### Insert data into clickhouse, 2 rows, 1 with sample weight at 1, the other with 99
  ```sql

4d5fc6ba4a26 :) insert into generic_metric_counters_raw_local FORMAT JSONColumns \
:-] {
      "use_case_id": ["custom"],
      "org_id": [1],
      "project_id": [1],
      "metric_id": [1],
      "timestamp": [1722378310],
      "retention_days": [90],
      "tags.key": [[1]],
      "tags.indexed_value": [[10]],
      "tags.raw_value": [["c_v_1"]],
      "set_values": [[]],
      "count_value": [1],
      "distribution_values": [[]],
      "metric_type": ["counter"],
      "materialization_version": [3],
      "timeseries_id": [123456],
      "partition": [0],
      "offset": [0],
      "granularities": [[2]],
      "sampling_weight": [1]
    }

INSERT INTO generic_metric_counters_raw_local FORMAT JSONColumns

Query id: 4006be39-0983-4ca7-aeda-c02cd05f4aa1

Ok.

4d5fc6ba4a26 :) insert into generic_metric_counters_raw_local FORMAT JSONColumns
                {
                  "use_case_id": ["custom"],
                  "org_id": [1],
                  "project_id": [1],
                  "metric_id": [1],
                  "timestamp": [1722378310],
                  "retention_days": [90],
                  "tags.key": [[1]],
                  "tags.indexed_value": [[10]],
                  "tags.raw_value": [["c_v_1"]],
                  "set_values": [[]],
                  "count_value": [1],
                  "distribution_values": [[]],
                  "metric_type": ["counter"],
                  "materialization_version": [3],
                  "timeseries_id": [123456],
                  "partition": [0],
                  "offset": [0],
                  "granularities": [[2]],
                  "sampling_weight": [99]
                }

INSERT INTO generic_metric_counters_raw_local FORMAT JSONColumns

Query id: 772c25b5-695e-49fa-9032-c605da0acee9

Ok.
  ```
  ### Query for unbiased values
```sql
SELECT sumMerge(value)
FROM generic_metric_counters_aggregated_local

Query id: 4bdfc292-ab21-468a-a2aa-57b305c02630

┌─sumMerge(value)─┐
│               2 │
└─────────────────┘

1 row in set. Elapsed: 0.005 sec.
```

### Query for biased values
```sql
SELECT sumMerge(value_weighted)
FROM generic_metric_counters_aggregated_local

Query id: a4bc6038-82fa-420e-b0c7-211976a90a98

┌─sumMerge(value_weighted)─┐
│                      100 │
└──────────────────────────┘

1 row in set. Elapsed: 0.005 sec.
```

  ### Insert a row with materialization version = 2
```sql
4d5fc6ba4a26 :) insert into generic_metric_counters_raw_local FORMAT JSONColumns
                {
                  "use_case_id": ["custom"],
                  "org_id": [0],
                  "project_id": [1],
                  "metric_id": [1],
                  "timestamp": [1722378310],
                  "retention_days": [90],
                  "tags.key": [[1]],
                  "tags.indexed_value": [[10]],
                  "tags.raw_value": [["c_v_1"]],
                  "set_values": [[]],
                  "count_value": [1],
                  "distribution_values": [[]],
                  "metric_type": ["counter"],
                  "materialization_version": [2],
                  "timeseries_id": [123456],
                  "partition": [0],
                  "offset": [0],
                  "granularities": [[2]],
                  "sampling_weight": [99]
                }

INSERT INTO generic_metric_counters_raw_local FORMAT JSONColumns

Query id: 7358e4ba-e5d9-4287-a40c-43b93670912b

Ok.

4d5fc6ba4a26 :) select SumMerge(value) from generic_metric_counters_aggregated_local where org_id = 0

SELECT SumMerge(value)
FROM generic_metric_counters_aggregated_local
WHERE org_id = 0

Query id: 6b50eaa2-b20a-41cb-aec2-53e2711d27fa

┌─SumMerge(value)─┐
│               1 │
└─────────────────┘

1 row in set. Elapsed: 0.006 sec.
```
</details>

### CODE REVIEW

File: snuba/snuba_migrations/generic_metrics/0052_counters_raw_add_sampling_weight.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    storage_set_key = StorageSetKey.GENERIC_METRICS_COUNTERS
+
+    local_table_name = "generic_metric_counters_raw_local"
+    dist_table_name = "generic_metric_counters_raw_dist"
+
+    columns: Sequence[tuple[Column[MigrationModifiers], str | None]] = [
+        (
+            Column(
+                "sampling_weight",
+                UInt(64, MigrationModifiers(default=str("1"))),
```
Comment (nikhars) [inline snuba/snuba_migrations/generic_metrics/0052_counters_raw_add_sampling_weight.py:22]: Let's add CODEC(ZSTD(1)) encoding to the column

File: snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    storage_set_key = StorageSetKey.GENERIC_METRICS_COUNTERS
+
+    local_table_name = "generic_metric_counters_aggregated_local"
+    dist_table_name = "generic_metric_counters_aggregated_dist"
+
+    columns: Sequence[tuple[Column[MigrationModifiers], str | None]] = [
+        (
+            Column(
+                "value_weighted",
+                AggregateFunction("sum", [Float(64)]),
```
Comment (nikhars) [inline snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py:23]: Lets add CODEC(ZSTD(1)) encoding to the column

File: snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        ),
+    ]
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.AddColumn(
+                storage_set=self.storage_set_key,
+                table_name=table_name,
+                column=column,
+                after=after,
```
Comment (nikhars) [inline snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py:35]: Nit: You can skip the after field

File: snuba/snuba_migrations/generic_metrics/0052_counters_raw_add_sampling_weight.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        ),
+    ]
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.AddColumn(
+                storage_set=self.storage_set_key,
+                table_name=table_name,
+                column=column,
+                after=after,
```
Comment (nikhars) [inline snuba/snuba_migrations/generic_metrics/0052_counters_raw_add_sampling_weight.py:34]: You can skip the after field.

File: snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py (hunk 3 above)
Comment (john-z-yang) [reply snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py:39]: Do we not want the weighted_value to be next to the regular value?

File: snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py (hunk 2 above)
Comment (john-z-yang) [reply snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py]: Done, thanks!

File: snuba/snuba_migrations/generic_metrics/0052_counters_raw_add_sampling_weight.py (hunk 1 above)
Comment (john-z-yang) [reply snuba/snuba_migrations/generic_metrics/0052_counters_raw_add_sampling_weight.py]: Fixed!

File: snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py (hunk 3 above)
Comment (nikhars) [reply snuba/snuba_migrations/generic_metrics/0053_counters_aggregated_add_sampling_weight.py:39]: It does make sense to add value_weighted after the value field in this case. Ignore my comment.

**Review by nikhars — APPROVED** [review]

**Review by enochtangg — APPROVED** [review]

**Pull request merged**